File: src/p1171/python/solution.py
class Solution:
    def removeZeroSumSublists(self, head: ListNode) -> ListNode:
        sentinel = ListNode(0)
        sentinel.next = head
        last = {}
        total, p = 0, sentinel
        while p:
            total += p.val
            last[total] = p
            p = p.next
        total, p = 0, sentinel
        while p:
            total += p.val
            p.next = last[total].next
            p = p.next
        return sentinel.next

File: src/p1171/cpp/solution.cpp
class Solution {
public:
    ListNode *removeZeroSumSublists(ListNode *head) {
        ListNode sentinel(0);
        sentinel.next = head;
        unordered_map<int, ListNode *> last;
        int total = 0;
        for (auto p = &sentinel; p; p = p->next) {
            total += p->val;
            last[total] = p;
        }
        total = 0;
        for (auto p = &sentinel; p; p = p->next) {
            total += p->val;
            p->next = last[total]->next;
        }
        return sentinel.next;
    }
};
